### ISSUE TITLE
Being able to use Expr.arr.slice() with expressions instead of fixed offset and length

### ISSUE THREAD

**Issue by sm-Fifteen** [pola-rs/polars#5166]
### Problem description

I'm dealing with a data format that's stored in a fixed length schema of 100 data columns (plus a few index columns), where each row has a "length", past which the data columns are all empty. The length may vary throughout the file on a per row basis. I'm trying to use `concat_list()` to merge all these columns together into a single one, and then slice the arrays based on the length specified for each row.

Something like this:

```py
ALL_100_DATA_COLS = pl.col(r"^data\d+$")

large_df.select([
    pl.col(["entity_name", "start_time", "duration", "row_length"]),
    pl.concat_list(ALL_100_DATA_COLS).arr.head( pl.col("row_length") ),
])
```

However, `head(n)` merely calls `slice(0,n)`, and `slice()` doesn't accept expressions as parameters, only fixed int values.

What I currently find myself having to do in Pandas to get acceptable performance (and I'm guessing my main alternative with Polars)  is to do a non-aggregating group-by on the value of `row_length` and use `apply()` to reshape and concatenate the contents on a "group" basis, but that's very error-prone with Pandas' auto-magic casting rules.

`pl.concat_list(ALL_100_DATA_COLS).apply( ... )` is also not much of an option, because I wouldn't be able to access the value of `row_length` from within the Python callback (on top of all the performance penalties that come with doing this, whether in Pandas or Polars).

It would be much more conveinient in my case to be able to simply slice using integer-valued columns.

**Comment by mcrumiller** [comment]
This sounds like a great improvement to the API, and one that I myself had to deal with in the past.

I would suggest the following, which isn't great:

1. Find all unique lengths via `df["row_length"].unique()`
2. Create a loop of expressions grabbing `arr.slice(offset, n)` for each unique n

So the code would look something like this:

```python
import polars as pl
from polars import col, when

lens = df["row_length"].unique()

data_col = col(r"^data\d+$")

row_ifs = [col("row_length") == len for len in lens]
slices = [col(r"^data\d+$").arr.slice(offset, len) for len in lens]

expr = when(row_ifs[0]).then(slices[0])
for row_if, slice in zip(row_ifs[1:], slices[1:]):
    expr = expr.when(row_if).then(slice)
expr = expr.otherwise('').alias("result")

df = df.with_column(expr)
```

**Comment by sm-Fifteen** [comment]
Tested with v0.14.20 on my end and it does solve the issue rather nicely (and *far* more elegantly than the Pandas groupby solution). Thank you @ritchie46!



> This sounds like a great improvement to the API, and one that I myself had to deal with in the past.
> 
> I would suggest the following

@mcrumiller: Forgot to reply, but I've been trying to use the lazy frames more and get used to the kind of workflow they bring, and that solution requires iterating over materialized data. It's a bit hard to follow, but it's the same general idea as the "groupby" thing I mentionned in the original post, where you basically split into "categories" based on length and then bulk-`slice()`.

I'll say, though, it's really cool to see that you can build large expressions to let polars handle this sort of request. I wasn't aware of `pl.when()`, since it isn't mentionned in the user guide and it's very easy to miss in the API reference, but having this sort of case-when ability is certainly going to come in handy elsewhere.
